aggregator/src/aggregator.rs

```diff
@@ -3459,6 +3633,222 @@ impl VdafOps {
+            .send((Box::new(report), rslt_tx))
+            .await
+            .unwrap();
+
+        // Await the result of writing the report.
+        // Unwrap safety: rslt_tx is always sent on before being dropped, and is never closed.
+        rslt_rx.await.unwrap()
+    }
+
+    async fn run_upload_batcher(
+        ds: Arc<Datastore<C>>,
+        mut report_rx: ReportWriteBatcherReceiver<C>,
+        max_batch_size: usize,
+        max_batch_write_delay: StdDuration,
+    ) {
+        let mut is_done = false;
+        let mut batch_expiry = TokioInstant::now();
+        let mut report_writers = Vec::with_capacity(max_batch_size);
+        let mut result_txs = Vec::with_capacity(max_batch_size);
+        while !is_done {
+            // Wait for an event of interest.
+            let write_batch = select! {
+                // Wait until we receive a report to be written (or the channel is closed due to the
+                // ReportWriteBatcher being dropped)...
+                item = report_rx.recv() => {
+                    match item {
+                        // We got an item. Add it to the current batch of reports to be written.
+                        Some((report_writer, rslt_tx)) => {
+                            if report_writers.is_empty() {
+                                batch_expiry = TokioInstant::now() + max_batch_write_delay;
+                            }
+                            report_writers.push(report_writer);
+                            result_txs.push(rslt_tx);
+                            report_writers.len() >= max_batch_size
+                        }
+
+                        // The channel is closed. Note this, and write any final reports that may be
+                        // batched before shutting down.
+                        None => {
+                            is_done = true;
+                            !report_writers.is_empty()
+                        },
+                    }
+                },
+
+                // ... or the current batch times out.
+                _ = sleep_until(batch_expiry), if !report_writers.is_empty() => true,
+            };
+
+            // If the event made us want to write the current batch to storage, do so.
+            if write_batch {
+                let ds = Arc::clone(&ds);
+                let result_writers =
+                    replace(&mut report_writers, Vec::with_capacity(max_batch_size));
+                let result_txs = replace(&mut result_txs, Vec::with_capacity(max_batch_size));
+                tokio::spawn(async move {
+                    Self::write_batch(ds, result_writers, result_txs).await;
+                });
+            }
+        }
+    }
+
+    async fn write_batch(
+        ds: Arc<Datastore<C>>,
+        report_writers: Vec<Box<dyn ReportWriter<C>>>,
+        result_txs: Vec<oneshot::Sender<Result<(), Arc<Error>>>>,
+    ) {
+        // Check preconditions.
+        assert_eq!(report_writers.len(), result_txs.len());
```

Comment:
Agreed, I was thinking that if `run_upload_batcher` just pushed tuples to one vector, that could simplify things.

Reply:
This idea doesn't work out because:
* I need to pass the report writers into the datastore transaction, so their container needs to be wrapped in an `Arc`.
* I need to consume the oneshot senders to `send` on them, so their container needs to _not_ be wrapped in an `Arc`.

That is, to make that idea work out, I would need to re-split the combined vector into two separate vectors before processing. I think I'd prefer the assert.